### ISSUE TITLE
Concurrency Issue in URI.LazyFragmentInitializer

### ISSUE THREAD

**Issue by ava-fred** [eclipse-emf/org.eclipse.emf#35]
We have a multi-threaded application using EMF, and we observed that when we increase the number of threads, there is a significant amount of contention on the lock used by the ReferenceQueue used by org.eclipse.emf.common.util.URI.LazyFragmentInitializer (among others).
We know we can avoid the use of this class by setting org.eclipse.emf.common.util.ReferenceClearingQueue=false, but we found that the time saved by avoiding the contention is lost elsewhere.
Is it possible to replace the ReferenceQueue by a queue implementation which doesn't need locking such as java.util.concurrent.ConcurrentLinkedQueue?

**Comment by ava-fred** [comment]
On further investigation I found out that the answer to my question is a resounding NO. The best we can hope for is that turning off the use of the ReferenceQueue for LazyFragmentInitializer only improves things for us. If it does, I will submit a PR implementing this.

**Comment by merks** [comment]
I'm not sure how you are measuring.  I generally use YourKit.  It supports both tracing and and sampling mode.  Be very cautious about how you interpret the numbers from sampling measurements. I personally don't trust them and from past experience folks have reported problems based on such sample measurements that really don't correspond to actual performance problems.

**Comment by ava-fred** [comment]
Just to address your point about measurements: our application is not interactive, it is more like a batch process. It does process individual requests, but those do not vary (much), so I can measure processing times of those and the overall runtime of the process, I do not rely on numbers from YourKit or JFR. I did use JFR to find the lock contention issue this is about.

While working on possible improvements for us I spotted something: org.eclipse.emf.common.util.URI.LazyFragmentInitializer.clear() looks like this:
```    @Override
    public void clear()
    {
      URI.Fragment uri = get();
      if (uri != null)
      {
        uri.fragment = splitInternFragment(fragment);
        uri.hashCode = ((uri.uri.hashCode * 31) + FRAGMENT_SEPARATOR) * CommonUtil.powerOf31(fragment.length()) + uri.fragment.hashCode();
      }
    }
```
My understanding of Java references has its limits, but IMO the get() call will always return null because of the call to enqueue() in the constructor of LazyFragmentInitializer. If I run the URITest unit test in the Eclipse Debugger, a breakpoint inside the if statement is never hit. I have also run the entire test suite in the build of EMF with "mined" code which would have crashed if the code inside the if was executed, nothing happened.
If you agree, then I will submit a PR which removes LazyFragmentInitializer completely.

**Comment by merks** [comment]
Reviewing it again, it does seem never to actually kick in.  ☹️

**Comment by ava-fred** [comment]
PR is submitted: https://github.com/eclipse-emf/org.eclipse.emf/pull/40.

**Comment by ava-fred** [comment]
Hi @merks , thanks for merging the PR. I'm guessing it will be included in EMF 2.36, can you give me an idea when that will be available?

**Comment by merks** [comment]
The builds are here:

https://download.eclipse.org/modeling/emf/emf/builds/

The next release will be 2.39.0

I was so organized for this cycle that I already created the release record:

https://projects.eclipse.org/projects/modeling.emf.emf/releases/2.39.0

EMF always follows the release train schedule:

- https://github.com/eclipse-simrel/.github/blob/main/wiki/SimRel/2024-09.md
- https://github.com/eclipse-simrel/.github/tree/main/wiki/SimRel

Pragmatically, because EMF is consumed by the Platform, it releases ahead of schedule typically by almost two weeks.

I've kicked off a nightly build and will also do a milestone build afterwards:

https://ci.eclipse.org/emf/job/build/job/master/
